### PR TITLE
storage: wrap closing chunk stored chan in mutex

### LINKED ISSUE #341: [Swarm] "panic: close of closed channel" when processing PSS messages
Hi there,

I've bumped into the following panic shortly after running a Swarm node.

#### System information

Geth version: 
```
Geth
Version: 1.8.3-unstable
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.10
Operating System: linux
GOPATH=/home/ubuntu/go
GOROOT=/usr/lib/go-1.10
```
Commit hash : `166c3e7e5be09e1452df92137ca2fd1b7de71bab` (swarm-network-rewrite branch)

#### Expected behaviour
Swarm node runs without crashing.

#### Actual behaviour
Swarm node crashes shortly after being run

#### Steps to reproduce the behaviour
1. Run a Swarm Node with PSS enabled 
2. Connect other nodes to the node
3. Starting sending PSS messages through the network

#### Backtrace

````
panic: close of closed channel

goroutine 35 [running]:
github.com/ethereum/go-ethereum/swarm/storage.(*LDBStore).Put(0xc4202bc000, 0xc422ef3620)
	/home/ubuntu/go/src/github.com/ethereum/go-ethereum/swarm/storage/ldbstore.go:567 +0x3be
github.com/ethereum/go-ethereum/swarm/storage.(*LocalStore).Put(0xc420079220, 0xc422c36120)
	/home/ubuntu/go/src/github.com/ethereum/go-ethereum/swarm/storage/localstore.go:101 +0x244
github.com/ethereum/go-ethereum/swarm/storage.(*DBAPI).Put(0xc42053a6b0, 0xc422c36120)
	/home/ubuntu/go/src/github.com/ethereum/go-ethereum/swarm/storage/dbapi.go:51 +0x39
github.com/ethereum/go-ethereum/swarm/network/stream.(*Delivery).processReceivedChunks(0xc4201914d0)
	/home/ubuntu/go/src/github.com/ethereum/go-ethereum/swarm/network/stream/delivery.go:209 +0x207
created by github.com/ethereum/go-ethereum/swarm/network/stream.NewDelivery
	/home/ubuntu/go/src/github.com/ethereum/go-ethereum/swarm/network/stream/delivery.go:50 +0xbe
````

### PR DESCRIPTION
**Pull request by samuil**
I did some testing to figure out what is root cause of the issue.

Chunks store in their state information if data was actually persisted (using `dbStored` chan). Operations on this state are protected by `ldbstore` mutex, but `ldbstore` has multiple instances created in code (i.e. `delivery` and `dpa`), and they don't share state between them.

Chunks are stored in `memstore` and they are indexed with their respective keys. When the message comes from two different sources, which is possible with PSS routing, first one might be stored in `memstore` already, and the second one will use cached chunk object  from the `memstore`.

The two concurrent processes try to persist chunk and update it's state at the same time, which on rare occasion will end up two instances of `ldbstore` working on the same chunk in memory.

This PR fixes #341

### CODE REVIEW

**Comment by nonsense** [comment]
lgtm, but probably best to wait for a second review, before merging, considering how important this is.

**Review by nonsense — APPROVED** [review]

**Review by janos — APPROVED** [review]
LGTM. We may need to change how chunk event propagation is done, but this change is fine for the current code.

**Pull request merged**
